### ISSUE TITLE
Recursion interfering with rule matching?

### ISSUE THREAD

**Issue by jvoigtlaender** [ndmitchell/hlint#1058]
I have the following rule in a file:
```
> cat .\rule.yaml
- warn: {lhs: "[] ++ x", rhs: x, name: Evaluate}
```
I'm testing it on the following code:
```haskell
> cat .\test.hs
f n = [] ++ f (n+1)

g n = [] ++ h (n+1)

h = g
```
The rule should apply in both the `f` and `g` definitions. The only difference between these two definitions is that `f` is directly recursive while `g` is not.

Surprisingly, only one of the two possible matches seems to be found:
```
> hlint.exe --version
HLint v3.1.5, (C) Neil Mitchell 2006-2020
> hlint.exe --hint=rule.yaml .\test.hs
test.hs:3:7-19: Warning: Evaluate
Found:
  [] ++ h (n + 1)
Perhaps:
  h (n + 1)

1 hint
```

--------------------

As an aside, should the mentioned rule,
```yaml
- warn: {lhs: "[] ++ x", rhs: x, name: Evaluate}
```
possibly along with its counterpart
```yaml
- warn: {lhs: "x ++ []", rhs: x, name: Evaluate}
```
be added to one of the `default` or `teaching` groups?

**Comment by ndmitchell** [comment]
To address the aside first, yes, those should be added to the default group.

As to why they don't match, the problem is the `checkDefine` clause in https://github.com/ndmitchell/hlint/blob/bbef0048ec88afb534d14311e27ea1bd91f27bde/src/Hint/Match.hs#L231-L239. The idea is that if you define:

```
a == b = compare a b == EQ
```

That's a perfectly valid definition. But in general, `compare a b == EQ` should be `a == b`. But here that would be infinite recursion, which is bad. I wonder what the precise semantics of `checkDefine` are, and if they should be slightly different?

**Comment by jvoigtlaender** [comment]
Interesting aspect

I'll add the two rules in question in a pull request.

About the `checkDefine` stuff: I haven't thought through all possible corner cases (obviously, since I wasn't even aware of this aspect before), but superficially it seems to me that this (kind of inadvertant introduction of infinite recursion) could only happen in the context of overloading or when applying a rule to an expression on the right-hand side of an equation for a function that itself appears in the rule (in either `lhs` or `rhs` of the rule)? Where actually the "overloading case" is a special case of the second one?

So, in the right-hand side of 
```haskell
a == b = ...
```
one shouldn't be allowed to apply any rule that mentions `==`, but in the right-hand side of something like
```haskell
function_written_by_user_that_hlint_knows_nothing_about args = ...
```
all rules could be applied without fear?

---------------------------

For context, the actual occurrence (from which the above minimal example was derived) was in student code that wouldn't have led to an infinite recursion (differently than the `f n = f (n+1)` case. The code looked something like this:
```haskell
f :: [A] -> [B]
f [] = []
f (x:xs)
 | x == ... = ... : f xs
 | x == ... = [...] ++ f xs
 | x == ... = [...] ++ f xs
 | otherwise = [] ++ f xs
```

**Comment by ndmitchell** [comment]
It's more than just overloading, although that is common. Anyone who defines their own Prelude, or runs HLint on the Prelude itself, runs into these issues. Anyone who generates hints also hits it. E.g. HLint itself defines the hint for HLint of:

```
- error: {lhs: idea Warning, rhs: warn}
```

And indeed, the definition of `warn` is:

```
warn = idea Warning
```

And replacing that hint on the above definition would be a bad idea. We could do something like trying to unify or even alpha-equivalent match the result with the LHS, but that seems fiddly. I wondered if it was sufficient to checkDefine before substitution, but the tests said no, so at least there are tests for this code :)
